types/account.go

```diff
@@ -25,144 +21,228 @@ const (
 	}
 	bz, err := hex.DecodeString(address)
 	if err != nil {
 		return nil, err
 	}
-	return Address(bz), nil
+	return ValAddress(bz), nil
 }
 
-// create an Address from a string
-func GetAccAddressBech32(address string) (addr Address, err error) {
-	bz, err := GetFromBech32(address, Bech32PrefixAccAddr)
+// create a ValAddress from a bech32 string
+func GetValAddressBech32(address string) (addr ValAddress, err error) {
+	bz, err := GetFromBech32(address, Bech32PrefixValAddr)
 	if err != nil {
 		return nil, err
 	}
-	return Address(bz), nil
+	return ValAddress(bz), nil
 }
 
-// create an Address from a string, panics on error
-func MustGetAccAddressBech32(address string) (addr Address) {
-	addr, err := GetAccAddressBech32(address)
+// Marshal needed for protobuf compatibility
+func (bz ValAddress) Marshal() ([]byte, error) {
+	return bz, nil
+}
+
+// Unmarshal needed for protobuf compatibility
+func (bz *ValAddress) Unmarshal(data []byte) error {
+	*bz = data
+	return nil
+}
+
+// Marshals to JSON using Bech32
+func (bz ValAddress) MarshalJSON() ([]byte, error) {
+	s := bz.String()
+	jbz := make([]byte, len(s)+2)
+	jbz[0] = '"'
+	copy(jbz[1:], []byte(s))
+	jbz[len(jbz)-1] = '"'
+	return jbz, nil
+}
+
+// Unmarshals from JSON assuming Bech32 encoding
+func (bz *ValAddress) UnmarshalJSON(data []byte) error {
+	if len(data) < 2 || data[0] != '"' || data[len(data)-1] != '"' {
+		return fmt.Errorf("Invalid bech32 string: %s", data)
+	}
+
+	bz2, err := GetValAddressBech32(string(data[1 : len(data)-1]))
 	if err != nil {
-		panic(err)
+		return err
 	}
-	return addr
+	*bz = bz2
+	return nil
 }
 
-// create a Pubkey from a string
-func GetAccPubKeyBech32(address string) (pk crypto.PubKey, err error) {
-	bz, err := GetFromBech32(address, Bech32PrefixAccPub)
+// Allow it to fulfill various interfaces in light-client, etc...
+func (bz ValAddress) Bytes() []byte {
+	return bz
+}
+
+func (bz ValAddress) String() string {
+	bech32Addr, err := bech32.ConvertAndEncode(Bech32PrefixValAddr, bz.Bytes())
 	if err != nil {
-		return nil, err
+		panic(err)
 	}
+	return bech32Addr
+}
 
-	pk, err = crypto.PubKeyFromBytes(bz)
-	if err != nil {
-		return nil, err
+// For Printf / Sprintf, returns bech32 when using %s
+func (bz ValAddress) Format(s fmt.State, verb rune) {
+	switch verb {
+	case 's':
+		s.Write([]byte(fmt.Sprintf("%s", bz.String())))
+	case 'p':
+		s.Write([]byte(fmt.Sprintf("%p", bz)))
+	default:
+		s.Write([]byte(fmt.Sprintf("%X", []byte(bz))))
 	}
+}
 
-	return pk, nil
+// Bech32ifyAccPub takes AccountPubKey and returns the bech32 encoded string
+func Bech32ifyAccPub(pub crypto.PubKey) (string, error) {
+	return bech32.ConvertAndEncode(Bech32PrefixAccPub, pub.Bytes())
 }
 
-// create an Pubkey from a string, panics on error
-func MustGetAccPubKeyBech32(address string) (pk crypto.PubKey) {
-	pk, err := GetAccPubKeyBech32(address)
+// MustBech32ifyAccPub panics on bech32-encoding failure
+func MustBech32ifyAccPub(pub crypto.PubKey) string {
```

Comment:
Do we really want to panic here? Why not return an error upstream and allow the caller/client to panic if they so choose to (cleans up the nomenclature too)?

Reply:
I think the `MustXYZ` convention is useful; it avoids a lot of code duplication since we usually need to panic if this fails.

The caller can use `Bech32ifyAccPub` if they want to handle the error themselves.

Reply:
Mhmm, I see. Fair point! Thanks.